File: core/src/main/java/dev/bee/kanjianki/core/KanjiGameEngine.java
package dev.bee.kanjianki.core;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Random;

public final class KanjiGameEngine {
    private static final int MAX_CHOICES = 4;

    public enum GameMode {
        MEANING_POP("meaning_pop", "Meaning Pop", "Kanji -> meaning", "Choose the meaning."),
        READING_RUSH("reading_rush", "Reading Rush", "Word -> reading", "Choose the reading."),
        CONFUSABLE_CLASH("confusable_clash", "Confusable Clash", "Meaning -> kanji", "Choose the kanji.");

        public final String id;
        public final String title;
        public final String label;
        public final String prompt;

        GameMode(String id, String title, String label, String prompt) {
            this.id = id;
            this.title = title;
            this.label = label;
            this.prompt = prompt;
        }
    }

    public static final class GameQuestion {
        public final GameMode mode;
        public final String targetKanji;
        public final String prompt;
        public final String promptDetail;
        public final String correctAnswer;
        public final List<String> choices;
        public final String explanation;

        public GameQuestion(
                GameMode mode,
                String targetKanji,
                String prompt,
                String promptDetail,
                String correctAnswer,
                List<String> choices,
                String explanation
        ) {
            this.mode = mode;
            this.targetKanji = clean(targetKanji);
            this.prompt = clean(prompt);
            this.promptDetail = clean(promptDetail);
            this.correctAnswer = clean(correctAnswer);
            this.choices = Collections.unmodifiableList(new ArrayList<>(cleanList(choices)));
            this.explanation = clean(explanation);
        }

        public boolean isCorrect(String selectedAnswer) {
            return answerKey(correctAnswer).equals(answerKey(selectedAnswer));
        }
    }

    public List<GameMode> availableModes(
            List<Records.DashboardRow> rows,
            List<Records.KanjiInventoryItem> inventory,
            List<Records.SimilarKanjiPair> pairs
    ) {
        List<GameMode> out = new ArrayList<>();
        for (GameMode mode : GameMode.values()) {
            if (nextQuestion(mode, rows, inventory, pairs, new Random(0L)) != null) {
                out.add(mode);
            }
        }
        return out;
    }

    public GameQuestion nextQuestion(
            GameMode mode,
            List<Records.DashboardRow> rows,
            List<Records.KanjiInventoryItem> inventory,
            List<Records.SimilarKanjiPair> pairs,
            Random random
    ) {
        GameMode safeMode = mode == null ? GameMode.MEANING_POP : mode;
        Random safeRandom = random == null ? new Random() : random;
        List<GameCandidate> candidates = candidates(rows, inventory);
        return switch (safeMode) {
            case MEANING_POP -> meaningQuestion(candidates, safeRandom);
            case READING_RUSH -> readingQuestion(candidates, safeRandom);
            case CONFUSABLE_CLASH -> confusableQuestion(candidates, pairs, safeRandom);
        };
    }

    private static GameQuestion meaningQuestion(List<GameCandidate> candidates, Random random) {
        List<GameCandidate> targets = targets(candidates, GameCandidate::hasMeaning);
        if (targets.isEmpty()) {
            return null;
        }
        GameCandidate target = randomCandidate(targets, random);
        List<String> choices = answerChoices(candidates, target.meaning, candidate -> candidate.meaning, random);
        if (choices.size() < 2) {
            return null;
        }
        return new GameQuestion(
                GameMode.MEANING_POP,
                target.kanji,
                target.kanji,
                "Pick the meaning",
                target.meaning,
                choices,
                target.kanji + " = " + target.meaning
        );
    }

    private static GameQuestion readingQuestion(List<GameCandidate> candidates, Random random) {
        List<GameCandidate> targets = targets(candidates, GameCandidate::hasReading);
        if (targets.isEmpty()) {
            return null;
        }
        GameCandidate target = randomCandidate(targets, random);
        List<String> choices = answerChoices(candidates, target.reading, candidate -> candidate.reading, random);
        if (choices.size() < 2) {
            return null;
        }
        return new GameQuestion(
                GameMode.READING_RUSH,
                target.kanji,
                target.expression.isEmpty() ? target.kanji : target.expression,
                "Pick the reading for " + target.kanji,
                target.reading,
                choices,
                target.explanation()
        );
    }

    private static GameQuestion confusableQuestion(
            List<GameCandidate> candidates,
            List<Records.SimilarKanjiPair> pairs,
            Random random
    ) {
        Map<String, List<String>> neighbors = neighborMap(pairs);
        List<GameCandidate> targets = new ArrayList<>();
        for (GameCandidate candidate : targets(candidates, GameCandidate::hasMeaning)) {
            List<String> direct = neighbors.get(candidate.kanji);
            if (direct != null && !direct.isEmpty()) {
                targets.add(candidate);
            }
        }
        if (targets.isEmpty()) {
            return null;
        }
        GameCandidate target = randomCandidate(targets, random);
        LinkedHashSet<String> choices = new LinkedHashSet<>();
        choices.add(target.kanji);
        List<String> direct = new ArrayList<>(neighbors.getOrDefault(target.kanji, Collections.emptyList()));
        Collections.shuffle(direct, random);
        for (String kanji : direct) {
            choices.add(kanji);
            if (choices.size() >= MAX_CHOICES) {
                break;
            }
        }
        List<String> shuffled = new ArrayList<>(choices);
        if (shuffled.size() < 2) {
            return null;
        }
        Collections.shuffle(shuffled, random);
        return new GameQuestion(
                GameMode.CONFUSABLE_CLASH,
                target.kanji,
                "Which kanji means " + target.meaning + "?",
                "Watch the shape",
                target.kanji,
                shuffled,
                target.kanji + " = " + target.meaning
        );
    }

    private static List<GameCandidate> candidates(
            List<Records.DashboardRow> rows,
            List<Records.KanjiInventoryItem> inventory
    ) {
        LinkedHashMap<String, GameCandidate> byKanji = new LinkedHashMap<>();
        if (rows != null) {
            for (Records.DashboardRow row : rows) {
                GameCandidate candidate = GameCandidate.fromRow(row);
                if (candidate.hasAnyAnswer()) {
                    byKanji.put(candidate.kanji, candidate);
                }
            }
        }
        if (inventory != null) {
            for (Records.KanjiInventoryItem item : inventory) {
                GameCandidate candidate = GameCandidate.fromInventory(item);
                if (candidate.hasAnyAnswer() && !byKanji.containsKey(candidate.kanji)) {
                    byKanji.put(candidate.kanji, candidate);
                }
            }
        }
        return new ArrayList<>(byKanji.values());
    }

    private static List<GameCandidate> targets(List<GameCandidate> candidates, CandidateFilter filter) {
        List<GameCandidate> active = new ArrayList<>();
        List<GameCandidate> fallback = new ArrayList<>();
        for (GameCandidate candidate : candidates) {
            if (filter.include(candidate)) {
                if (candidate.fromDashboard) {
                    active.add(candidate);
                } else {
                    fallback.add(candidate);
                }
            }
        }
        return active.isEmpty() ? fallback : active;
    }

    private static GameCandidate randomCandidate(List<GameCandidate> candidates, Random random) {
        return candidates.get(random.nextInt(candidates.size()));
    }

    private static List<String> answerChoices(
            List<GameCandidate> candidates,
            String correct,
            AnswerExtractor extractor,
            Random random
    ) {
        LinkedHashSet<String> answers = new LinkedHashSet<>();
        String safeCorrect = clean(correct);
        answers.add(safeCorrect);
        List<String> decoys = new ArrayList<>();
        for (GameCandidate candidate : candidates) {
            String answer = clean(extractor.answer(candidate));
            if (!answer.isEmpty() && !answerKey(answer).equals(answerKey(safeCorrect))) {
                decoys.add(answer);
            }
        }
        decoys = uniqueAnswers(decoys);
        Collections.shuffle(decoys, random);
        for (String decoy : decoys) {
            answers.add(decoy);
            if (answers.size() >= MAX_CHOICES) {
                break;
            }
        }
        List<String> shuffled = new ArrayList<>(answers);
        Collections.shuffle(shuffled, random);
        return shuffled;
    }

    private static List<String> uniqueAnswers(List<String> values) {
        LinkedHashMap<String, String> byKey = new LinkedHashMap<>();
        for (String value : values) {
            byKey.putIfAbsent(answerKey(value), value);
        }
        return new ArrayList<>(byKey.values());
    }

    private static Map<String, List<String>> neighborMap(List<Records.SimilarKanjiPair> pairs) {
        Map<String, List<String>> out = new HashMap<>();
        if (pairs == null) {
            return out;
        }
        for (Records.SimilarKanjiPair pair : pairs) {
            if (pair == null) {
                continue;
            }
            String a = clean(pair.kanjiA);
            String b = clean(pair.kanjiB);
            if (a.isEmpty() || b.isEmpty() || a.equals(b)) {
                continue;
            }
            addNeighbor(out, a, b);
            addNeighbor(out, b, a);
        }
        return out;
    }

    private static void addNeighbor(Map<String, List<String>> neighbors, String kanji, String neighbor) {
        List<String> direct = neighbors.computeIfAbsent(kanji, ignored -> new ArrayList<>());
        if (!direct.contains(neighbor)) {
            direct.add(neighbor);
        }
    }

    private static List<String> cleanList(List<String> values) {
        List<String> out = new ArrayList<>();
        if (values == null) {
            return out;
        }
        for (String value : values) {
            String cleaned = clean(value);
            if (!cleaned.isEmpty()) {
                out.add(cleaned);
            }
        }
        return out;
    }

    private static String firstNonEmpty(String first, String second) {
        String safeFirst = clean(first);
        return safeFirst.isEmpty() ? clean(second) : safeFirst;
    }

    private static String clean(String value) {
        if (value == null) {
            return "";
        }
        return value.replace("Meaning:", "")
                .replaceAll("\\s+", " ")
                .trim();
    }

    private static String answerKey(String value) {
        return clean(value).toLowerCase(Locale.ROOT);
    }

    private interface CandidateFilter {
        boolean include(GameCandidate candidate);
    }

    private interface AnswerExtractor {
        String answer(GameCandidate candidate);
    }

    private static final class GameCandidate {
        final String kanji;
        final String meaning;
        final String reading;
        final String expression;
        final boolean fromDashboard;

        GameCandidate(String kanji, String meaning, String reading, String expression, boolean fromDashboard) {
            this.kanji = clean(kanji);
            this.meaning = clean(meaning);
            this.reading = clean(reading);
            this.expression = clean(expression);
            this.fromDashboard = fromDashboard;
        }

        static GameCandidate fromRow(Records.DashboardRow row) {
            if (row == null) {
                return new GameCandidate("", "", "", "", true);
            }
            String exampleMeaning = "";
            String exampleReading = "";
            String expression = "";
            for (Records.Example example : row.examples) {
                if (exampleMeaning.isEmpty()) {
                    exampleMeaning = example.meaning;
                }
                if (exampleReading.isEmpty()) {
                    exampleReading = example.reading;
                }
                if (expression.isEmpty()) {
                    expression = example.expression;
                }
            }
            return new GameCandidate(
                    row.kanji,
                    firstNonEmpty(row.primaryMeaning, exampleMeaning),
                    firstNonEmpty(row.reading, exampleReading),
                    firstNonEmpty(expression, row.kanji),
                    true
            );
        }

        static GameCandidate fromInventory(Records.KanjiInventoryItem item) {
            if (item == null) {
                return new GameCandidate("", "", "", "", false);
            }
            return new GameCandidate(item.kanji, item.primaryMeaning, item.readings, item.kanji, false);
        }

        boolean hasAnyAnswer() {
            return !kanji.isEmpty() && (hasMeaning() || hasReading());
        }

        boolean hasMeaning() {
            return !meaning.isEmpty();
        }

        boolean hasReading() {
            return !reading.isEmpty();
        }

        String explanation() {
            if (meaning.isEmpty()) {
                return kanji + " = " + reading;
            }
            return kanji + " = " + reading + " · " + meaning;
        }
    }
}

File: core/src/main/kotlin/dev/bee/kanjianki/core/KanjiGameEngine.kt
package dev.bee.kanjianki.core

import java.security.SecureRandom
import java.util.Collections
import java.util.Locale
import java.util.Random

class KanjiGameEngine {
    enum class GameMode(
        @JvmField val id: String,
        @JvmField val title: String,
        @JvmField val label: String,
        @JvmField val prompt: String,
        @JvmField val lockedLabel: String = "",
    ) {
        MEANING_POP("meaning_pop", "Meaning Pop", "Kanji -> meaning", "Choose the meaning."),
        READING_RUSH("reading_rush", "Reading Rush", "Word -> reading", "Choose the reading."),
        CONFUSABLE_CLASH("confusable_clash", "Confusable Clash", "Meaning -> kanji", "Choose the kanji."),
        MISS_SWEEP("miss_sweep", "Miss Sweep", "Recent misses", "Choose the meaning.", "Need 2+ recent misses"),
    }

    class GameQuestion(
        @JvmField val mode: GameMode,
        targetKanji: String?,
        prompt: String?,
        promptDetail: String?,
        correctAnswer: String?,
        choices: List<String?>?,
        explanation: String?,
    ) {
        @JvmField val targetKanji: String = clean(targetKanji)
        @JvmField val prompt: String = clean(prompt)
        @JvmField val promptDetail: String = clean(promptDetail)
        @JvmField val correctAnswer: String = clean(correctAnswer)
        @JvmField val choices: List<String> = Collections.unmodifiableList(ArrayList(cleanChoices(choices)))
        @JvmField val explanation: String = clean(explanation)

        fun isCorrect(selectedAnswer: String?): Boolean {
            return answerKey(correctAnswer) == answerKey(selectedAnswer)
        }
    }

    fun availableModes(
        rows: List<RecordsImportModels.DashboardRow?>?,
        inventory: List<RecordsImportModels.KanjiInventoryItem?>?,
        pairs: List<RecordsImportModels.SimilarKanjiPair?>?,
        recentMissKanji: List<String?>? = null,
    ): List<GameMode> {
        val out = ArrayList<GameMode>()
        for (mode in GameMode.values()) {
            if (nextQuestion(mode, rows, inventory, pairs, SecureRandom(), recentMissKanji) != null) {
                out.add(mode)
            }
        }
        return out
    }

    fun nextQuestion(
        mode: GameMode?,
        rows: List<RecordsImportModels.DashboardRow?>?,
        inventory: List<RecordsImportModels.KanjiInventoryItem?>?,
        pairs: List<RecordsImportModels.SimilarKanjiPair?>?,
        random: Random?,
        recentMissKanji: List<String?>? = null,
    ): GameQuestion? {
        val safeMode = mode ?: GameMode.MEANING_POP
        val safeRandom = random ?: SecureRandom()
        val candidates = candidates(rows, inventory)
        return when (safeMode) {
            GameMode.MEANING_POP -> meaningQuestion(candidates, safeRandom)
            GameMode.READING_RUSH -> readingQuestion(candidates, safeRandom)
            GameMode.CONFUSABLE_CLASH -> confusableQuestion(candidates, pairs, safeRandom)
            GameMode.MISS_SWEEP -> missSweepQuestion(candidates, recentMissKanji, safeRandom)
        }
    }

    private data class GameCandidate(
        val kanji: String,
        val meaning: String,
        val reading: String,
        val expression: String,
        val fromDashboard: Boolean,
    ) {
        fun hasAnyAnswer(): Boolean = kanji.isNotEmpty() && (hasMeaning() || hasReading())

        fun hasMeaning(): Boolean = meaning.isNotEmpty()

        fun hasReading(): Boolean = reading.isNotEmpty()

        fun explanation(): String {
            if (meaning.isEmpty()) {
                return "$kanji = $reading"
            }
            return "$kanji = $reading · $meaning"
        }

        companion object {
            fun fromRow(row: RecordsImportModels.DashboardRow?): GameCandidate {
                if (row == null) {
                    return GameCandidate("", "", "", "", true)
                }
                var exampleMeaning = ""
                var exampleReading = ""
                var expression = ""
                for (example in row.examples) {
                    if (exampleMeaning.isEmpty()) {
                        exampleMeaning = example.meaning
                    }
                    if (expression.isEmpty() &&
                        example.expression.isNotBlank() &&
                        example.reading.isNotBlank()
                    ) {
                        expression = example.expression
                        exampleReading = example.reading
                    }
                }
                return GameCandidate(
                    clean(row.kanji),
                    firstNonEmpty(row.primaryMeaning, exampleMeaning),
                    if (expression.isEmpty()) {
                        firstNonEmpty(row.reading, exampleReading)
                    } else {
                        clean(exampleReading)
                    },
                    firstNonEmpty(expression, row.kanji),
                    true,
                )
            }

            fun fromInventory(item: RecordsImportModels.KanjiInventoryItem?): GameCandidate {
                if (item == null) {
                    return GameCandidate("", "", "", "", false)
                }
                return GameCandidate(
                    clean(item.kanji),
                    clean(item.primaryMeaning),
                    clean(item.readings),
                    clean(item.kanji),
                    false,
                )
            }

            private fun firstNonEmpty(first: String?, second: String?): String {
                val safeFirst = clean(first)
                return if (safeFirst.isEmpty()) clean(second) else safeFirst
            }
        }
    }

    companion object {
        private const val MAX_CHOICES = 4

        private fun meaningQuestion(candidates: List<GameCandidate>, random: Random): GameQuestion? {
            val targets = targets(candidates) { it.hasMeaning() }
            if (targets.isEmpty()) {
                return null
            }
            val target = randomCandidate(targets, random)
            val choices = answerChoices(candidates, target.meaning, { it.meaning }, random)
            if (choices.size < 2) {
                return null
            }
            return GameQuestion(
                GameMode.MEANING_POP,
                target.kanji,
                target.kanji,
                "Pick the meaning",
                target.meaning,
                choices,
                "${target.kanji} = ${target.meaning}",
            )
        }

        private fun readingQuestion(candidates: List<GameCandidate>, random: Random): GameQuestion? {
            val targets = targets(candidates) { it.hasReading() }
            if (targets.isEmpty()) {
                return null
            }
            val target = randomCandidate(targets, random)
            val choices = answerChoices(candidates, target.reading, { it.reading }, random)
            if (choices.size < 2) {
                return null
            }
            return GameQuestion(
                GameMode.READING_RUSH,
                target.kanji,
                if (target.expression.isEmpty()) target.kanji else target.expression,
                "Pick the reading for ${target.kanji}",
                target.reading,
                choices,
                target.explanation(),
            )
        }

        private fun confusableQuestion(
            candidates: List<GameCandidate>,
            pairs: List<RecordsImportModels.SimilarKanjiPair?>?,
            random: Random,
        ): GameQuestion? {
            val neighbors = neighborMap(pairs)
            val targets = ArrayList<GameCandidate>()
            for (candidate in targets(candidates) { it.hasMeaning() }) {
                val direct = neighbors[candidate.kanji]
                if (!direct.isNullOrEmpty()) {
                    targets.add(candidate)
                }
            }
            if (targets.isEmpty()) {
                return null
            }
            val target = randomCandidate(targets, random)
            val choices = LinkedHashSet<String>()
            choices.add(target.kanji)
            val direct = ArrayList(neighbors.getOrDefault(target.kanji, emptyList()))
            Collections.shuffle(direct, random)
            for (kanji in direct) {
                choices.add(kanji)
                if (choices.size >= MAX_CHOICES) {
                    break
                }
            }
            val shuffled = ArrayList(choices)
            if (shuffled.size < 2) {
                return null
            }
            Collections.shuffle(shuffled, random)
            return GameQuestion(
                GameMode.CONFUSABLE_CLASH,
                target.kanji,
                "Which kanji means ${target.meaning}?",
                "Watch the shape",
                target.kanji,
                shuffled,
                "${target.kanji} = ${target.meaning}",
            )
        }

        private fun missSweepQuestion(
            candidates: List<GameCandidate>,
            recentMissKanji: List<String?>?,
            random: Random,
        ): GameQuestion? {
            val missSet = HashSet<String>()
            recentMissKanji.orEmpty().forEach { k ->
                val cleaned = clean(k)
                if (cleaned.isNotEmpty()) missSet.add(cleaned)
            }
            val missTargets = candidates.filter { it.hasMeaning() && missSet.contains(it.kanji) }
            if (missTargets.size < 2) return null
            val target = randomCandidate(missTargets, random)
            val choices = answerChoices(candidates, target.meaning, { it.meaning }, random)
            if (choices.size < 2) return null
            return GameQuestion(
                GameMode.MISS_SWEEP,
                target.kanji,
                target.kanji,
                "Pick the meaning",
                target.meaning,
                choices,
                "${target.kanji} = ${target.meaning}",
            )
        }

        private fun candidates(
            rows: List<RecordsImportModels.DashboardRow?>?,
            inventory: List<RecordsImportModels.KanjiInventoryItem?>?,
        ): List<GameCandidate> {
            val byKanji = LinkedHashMap<String, GameCandidate>()
            if (rows != null) {
                for (row in rows) {
                    val candidate = GameCandidate.fromRow(row)
                    if (candidate.hasAnyAnswer()) {
                        byKanji[candidate.kanji] = candidate
                    }
                }
            }
            if (inventory != null) {
                for (item in inventory) {
                    val candidate = GameCandidate.fromInventory(item)
                    if (candidate.hasAnyAnswer() && !byKanji.containsKey(candidate.kanji)) {
                        byKanji[candidate.kanji] = candidate
                    }
                }
            }
            return ArrayList(byKanji.values)
        }

        private fun targets(
            candidates: List<GameCandidate>,
            filter: (GameCandidate) -> Boolean,
        ): List<GameCandidate> {
            val active = ArrayList<GameCandidate>()
            val fallback = ArrayList<GameCandidate>()
            for (candidate in candidates) {
                if (filter(candidate)) {
                    if (candidate.fromDashboard) {
                        active.add(candidate)
                    } else {
                        fallback.add(candidate)
                    }
                }
            }
            return if (active.isEmpty()) fallback else active
        }

        private fun randomCandidate(candidates: List<GameCandidate>, random: Random): GameCandidate {
            return candidates[random.nextInt(candidates.size)]
        }

        private fun answerChoices(
            candidates: List<GameCandidate>,
            correct: String?,
            extractor: (GameCandidate) -> String,
            random: Random,
        ): List<String> {
            val answers = LinkedHashSet<String>()
            val safeCorrect = clean(correct)
            answers.add(safeCorrect)
            var decoys = ArrayList<String>()
            for (candidate in candidates) {
                val answer = clean(extractor(candidate))
                if (answer.isNotEmpty() && answerKey(answer) != answerKey(safeCorrect)) {
                    decoys.add(answer)
                }
            }
            decoys = ArrayList(uniqueAnswers(decoys))
            Collections.shuffle(decoys, random)
            for (decoy in decoys) {
                answers.add(decoy)
                if (answers.size >= MAX_CHOICES) {
                    break
                }
            }
            val shuffled = ArrayList(answers)
            Collections.shuffle(shuffled, random)
            return shuffled
        }

        private fun uniqueAnswers(values: List<String>): List<String> {
            val byKey = LinkedHashMap<String, String>()
            for (value in values) {
                byKey.putIfAbsent(answerKey(value), value)
            }
            return ArrayList(byKey.values)
        }

        private fun neighborMap(pairs: List<RecordsImportModels.SimilarKanjiPair?>?): Map<String, List<String>> {
            val out = HashMap<String, MutableList<String>>()
            if (pairs == null) {
                return out
            }
            for (pair in pairs) {
                if (hasUsablePair(pair)) {
                    val a = clean(pair!!.kanjiA)
                    val b = clean(pair.kanjiB)
                    addNeighbor(out, a, b)
                    addNeighbor(out, b, a)
                }
            }
            return out
        }

        private fun hasUsablePair(pair: RecordsImportModels.SimilarKanjiPair?): Boolean {
            if (pair == null) {
                return false
            }
            val a = clean(pair.kanjiA)
            val b = clean(pair.kanjiB)
            return a.isNotEmpty() && b.isNotEmpty() && a != b
        }

        private fun addNeighbor(neighbors: MutableMap<String, MutableList<String>>, kanji: String, neighbor: String) {
            val direct = neighbors.computeIfAbsent(kanji) { ArrayList() }
            if (!direct.contains(neighbor)) {
                direct.add(neighbor)
            }
        }

        private fun cleanChoices(values: List<String?>?): List<String> {
            val out = ArrayList<String>()
            if (values == null) {
                return out
            }
            for (value in values) {
                val cleaned = clean(value)
                if (cleaned.isNotEmpty()) {
                    out.add(cleaned)
                }
            }
            return out
        }

        private fun clean(value: String?): String {
            if (value == null) {
                return ""
            }
            return value.replace("Meaning:", "")
                .replace(Regex("\\s+"), " ")
                .trim()
        }

        private fun answerKey(value: String?): String {
            return clean(value).lowercase(Locale.ROOT)
        }
    }
}
